spec/unit/solr_tools_spec.rb

```diff
@@ -15,20 +15,165 @@
+      allow_any_instance_of(@harvester_class).to receive(:ping_solr).and_return(false)
+      expect_any_instance_of(@harvester_class).to receive(:ping_solr)
+      allow_any_instance_of(@harvester_class).to receive(:ping_source).and_return(false)
+      expect_any_instance_of(@harvester_class).to receive(:ping_source)
+
+      @cli.options = { data_center: %w(nsidc), die_on_failure: false, environment: 'dev' }
+      expect { @cli.ping }.to raise_error(SystemExit) do |error|
+        expect(error.status).to eq(SearchSolrTools::Errors::HarvestError::ERRCODE_SOLR_PING + SearchSolrTools::Errors::HarvestError::ERRCODE_SOURCE_PING)
+      end
+    end
+  end
+
   describe '#harvest' do
+    let (:ingest_ok) { SearchSolrTools::Helpers::HarvestStatus::INGEST_OK }
+    let (:ingest_invalid_doc) { SearchSolrTools::Helpers::HarvestStatus::INGEST_ERR_INVALID_DOC }
+    let (:ingest_solr_err) { SearchSolrTools::Helpers::HarvestStatus::INGEST_ERR_SOLR_ERROR }
+    let (:doc_result) { { num_docs: 3, add_docs: ['doc1', 'doc2', 'doc3'], failure_ids: [] } }
+
+    before(:each) do
+      @harvester_class = SearchSolrTools::Harvesters::NsidcJson
+    end
+
     it 'calls the selected harvester classes' do
       puts 'CLI' + @cli.harvester_map.to_s
-      [SearchSolrTools::Harvesters::Adc, SearchSolrTools::Harvesters::Echo].each do |harvester_class|
-        allow_any_instance_of(harvester_class).to receive(:harvest_and_delete).and_return(true)
-        expect_any_instance_of(harvester_class).to receive(:harvest_and_delete)
+      [SearchSolrTools::Harvesters::Adc, SearchSolrTools::Harvesters::Echo].each do |test_harvester_class|
+        allow_any_instance_of(test_harvester_class).to receive(:ping_solr).and_return(true)
+        allow_any_instance_of(test_harvester_class).to receive(:ping_source).and_return(true)
+        allow_any_instance_of(test_harvester_class).to receive(:harvest_and_delete).and_return(true)
+        expect_any_instance_of(test_harvester_class).to receive(:harvest_and_delete)
       end
       @cli.options = { data_center: %w(echo adc), die_on_failure: false, environment: 'dev' }
       @cli.harvest
     end
 
+    it 'fails when an invalid datacenter is provided' do
+      @cli.options = { data_center: ['not a real datacenter'], environment: 'dev' }
+      expect { @cli.harvest }.to raise_exception(SystemExit) do |error|
+        expect(error.status).to eql(SearchSolrTools::Errors::HarvestError::ERRCODE_OTHER)
+      end
+    end
+
+    # TODO: Not testing what it thinks it's testing! The test will pass even if
+    #  die_on_failure is set to false. die_on_failure is set in response to DCS errors,
+    #  not non-existing data center errors.
+    #  NOTE - I copied the main test logic above, as it's really testing to see if there is an
+    #  invalid datacenter.  If we want to keep the "die_on_failure", we need to change the
+    #  test logic here.
     it 'fails on failure if die_on_failure is true' do
```

Comment:
These comments are confusing. Lines 99-101 indicate that the test doesn't actually work, since `die_on_failure` actually refers to handling of DCS (harvesting) errors. What "main test logic" is line 102 referring to? In order to actually test `die_on_failure`, DCS (harvesting) needs to be stubbed to fail. As it exists now, this test simply replicates the test on line 92. If you want to prove that `die_on_failure` has no bearing on the handling of invalid data centers, add another test in which `die_on_failure` is set to false. The results should be the same as for this test.

Reply:
In other words, it appears to me that the comments in lines 102-104 don't add anything to the comments already in lines 99-101.

Reply:
Perhaps "I copied the main test logic above" means that you created the test on lines 92-97 by copying lines 105-108?

Reply:
I may just remove this test.  The idea of die_on_failure may be obsolete now anyway, since we are wrapping the failures in a status object, and will return an "error" exit code if this happens anyway.